frontend/src/components/navbar/UserMenu.tsx

```diff
@@ -1,23 +1,84 @@
 import Cookies from 'js-cookie';
 import { useNavigate } from 'react-router-dom';
-import { Button, Menu } from '@mantine/core';
+import { Menu, Avatar } from '@mantine/core';
+import authFetch from '../../rmoods/client/authFetch.ts';
+import { useEffect, useState } from 'react';
+import { jwtDecode } from 'jwt-decode';
+import { JwtClaims } from '../../rmoods/jwt.ts';
+import { changeImageSize } from '../../utility/changeImageSize.ts';
 
+/**
+ * User interface representing the user data.
+ */
+interface User {
+  name: string;
+  picture: string;
+}
+
+/**
+ * UserMenu component that displays the user menu.
+ * @returns {JSX.Element} The UserMenu component.
+ */
 const UserMenu = () => {
   const navigate = useNavigate();
+  const [user, setUser] = useState<User | null>(null);
+  const [loading, setLoading] = useState(true);
+
+  useEffect(() => {
+    const token = Cookies.get('RMOODS_JWT');
+    if (!token) {
+      navigate('/login');
+      return;
+    }
+
+    let isMounted = true;
+    try {
+      const data = jwtDecode<JwtClaims>(token);
+      const id = data.userInfo.id;
+
+      authFetch('http://localhost:8001/api/user?id=' + id, token).then(
+        (response) => {
+          response.json().then((data) => {
+            if (isMounted) {
+              setUser(data);
+              setLoading(false);
+            }
+          });
+        }
+      );
+    } catch (error) {
+      console.error('JWT token could not be decoded.', error);
+      if (isMounted) {
+        setLoading(false);
+      }
+    }
+    return () => {
+      isMounted = false;
+    };
+  }, [navigate]);
```

Comment:
This will have to be eventually rewritten to use TanStack Query, but no issues here.